File: sdk/src/main/java/com/cxense/cxensesdk/model/WidgetContext.java
package com.cxense.cxensesdk.model;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Context for Widget. It is required for loading items.
 *
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2017-06-05).
 */
@SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
public final class WidgetContext {
    private String url;
    private String pageclass;
    private String sentiment;
    private boolean recommending;
    private Map<String, String> categories;
    private List<String> keywords;
    private List<String> neighbors;
    private String referrer;
    private List<ContextParameter> parameters;

    public WidgetContext() {
    }

    /**
     * Initialize Widget context from {@link WidgetContext.Builder}.
     *
     * @param builder builder instance.
     */
    public WidgetContext(@NonNull Builder builder) {
        this();
        url = builder.url;
        pageclass = builder.pageclass;
        sentiment = builder.sentiment;
        recommending = builder.recommending;
        categories = Collections.unmodifiableMap(builder.categories);
        keywords = Collections.unmodifiableList(builder.keywords);
        neighbors = Collections.unmodifiableList(builder.neighbors);
        referrer = builder.referrer;
        parameters = Collections.unmodifiableList(builder.parameters);
    }

    /**
     * Gets the URL for context.
     *
     * @return URL.
     */
    @Nullable
    public String getUrl() {
        return url;
    }

    /**
     * Gets the pageclass of the current page.
     *
     * @return the pageclass of the current page.
     */
    @Nullable
    public String getPageclass() {
        return pageclass;
    }

    /**
     * Gets the sentiment of the current page.
     *
     * @return specified sentiment.
     */
    @Nullable
    public String getSentiment() {
        return sentiment;
    }

    /**
     * Gets the recs-recommending setting of the current page.
     *
     * @return recs-recommending setting.
     */
    public boolean isRecommending() {
        return recommending;
    }

    /**
     * Gets map for categories of the current page.
     *
     * @return unmodifiable map for categories.
     */
    @NonNull
    public Map<String, String> getCategories() {
        return categories;
    }

    /**
     * Gets list of keywords describing the context.
     *
     * @return unmodifiable list of keywords
     */
    @NonNull
    public List<String> getKeywords() {
        return keywords;
    }

    /**
     * Gets list of article IDs already linked to near this widget.
     * By default we will try to avoid recommending the same articles.
     *
     * @return unmodifiable list of linked articles IDs.
     */
    @NonNull
    public List<String> getNeighbors() {
        return neighbors;
    }

    /**
     * Gets the URI of the document that linked to this page.
     *
     * @return referrer URI.
     */
    @Nullable
    public String getReferrer() {
        return referrer;
    }

    /**
     * Gets unmodifiable list of {@link ContextParameter} objects.
     * This is used for placeholders.
     *
     * @return Unmodifiable list of parameters.
     */
    @NonNull
    public List<ContextParameter> getParameters() {
        return parameters;
    }

    public static class Builder {
        String url;
        String pageclass;
        String sentiment;
        boolean recommending;
        Map<String, String> categories = new HashMap<>();
        List<String> keywords = new ArrayList<>();
        List<String> neighbors = new ArrayList<>();
        String referrer;
        List<ContextParameter> parameters = new ArrayList<>();

        /**
         * Initialize builder
         *
         * @param url url for widget
         */
        public Builder(@Nullable String url) {
            this.url = url;
        }

        /**
         * Sets URL for widget
         *
         * @param url new URL
         * @return {@link Builder} instance
         */
        @NonNull
        public WidgetContext.Builder setUrl(@Nullable String url) {
            this.url = url;
            return this;
        }

        /**
         * Set pageclass for widget.
         *
         * @param pageclass the pageclass of the current page.
         * @return {@link Builder} instance
         */
        @NonNull
        public WidgetContext.Builder setPageclass(@Nullable String pageclass) {
            this.pageclass = pageclass;
            return this;
        }

        /**
         * Set sentiment for widget.
         *
         * @param sentiment the sentiment of the current page.
         * @return {@link Builder} instance
         */
        @NonNull
        public WidgetContext.Builder setSentiment(@Nullable String sentiment) {
            this.sentiment = sentiment;
            return this;
        }

        /**
         * Set recs-recommending setting for widget.
         *
         * @param recommending recs-recommending setting of the current page.
         * @return {@link Builder} instance
         */
        @NonNull
        public WidgetContext.Builder setRecommending(@Nullable boolean recommending) {
            this.recommending = recommending;
            return this;
        }

        /**
         * Set categories for widget.
         *
         * @param categories map for categories of the current page.
         * @return {@link Builder} instance
         */
        @NonNull
        public WidgetContext.Builder setCategories(@NonNull Map<String, String> categories) {
            this.categories = new HashMap<>(categories);
            return this;
        }

        /**
         * Set list of keywords for widget.
         *
         * @param keywords list of keywords describing the context.
         * @return {@link Builder} instance
         */
        @NonNull
        public WidgetContext.Builder setKeywords(@NonNull Collection<String> keywords) {
            this.keywords = new ArrayList<>(keywords);
            return this;
        }

        /**
         * Set list of article IDs for widget.
         *
         * @param neighbors list of linked articles IDs.
         * @return {@link Builder} instance
         */
        @NonNull
        public WidgetContext.Builder setNeighbors(@NonNull Collection<String> neighbors) {
            this.neighbors = new ArrayList<>(neighbors);
            return this;
        }

        /**
         * Set referrer for widget.
         *
         * @param referrer referrer URI.
         * @return {@link Builder} instance
         */
        @NonNull
        public WidgetContext.Builder setReferrer(@Nullable String referrer) {
            this.referrer = referrer;
            return this;
        }

        /**
         * Set parameters for widget.
         * This is used for placeholders.
         *
         * @param parameters list of {@link ContextParameter} objects.
         * @return {@link Builder} instance
         */
        @NonNull
        public WidgetContext.Builder setParameters(@NonNull Collection<ContextParameter> parameters) {
            this.parameters = new ArrayList<>(parameters);
            return this;
        }

        /**
         * Build widget context
         *
         * @return {@link WidgetContext} instance
         */
        @NonNull
        public WidgetContext build() {
            return new WidgetContext(this);
        }
    }
}

File: sdk/src/main/java/io/piano/android/cxense/model/WidgetContext.kt
package io.piano.android.cxense.model

import com.squareup.moshi.JsonClass
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import java.util.Collections

/**
 * Provides information about the current context / page for Widget
 * @property url url for widget, required.
 * @property referrer referrer url.
 * @property pageclass the pageclass of the current page.
 * @property sentiment the sentiment of the current page.
 * @property recommending recs-recommending setting of the current page.
 * @property categories map for categories of the current page.
 * @property keywords list of keywords describing the context.
 * @property neighbors list of linked articles IDs.
 * @property parameters list of [ContextParameter] objects.
 */
@Suppress("unused", "MemberVisibilityCanBePrivate") // Public API.
@JsonClass(generateAdapter = true)
class WidgetContext internal constructor(
    val url: String,
    val referrer: String?,
    val pageclass: String?,
    val sentiment: String?,
    val recommending: Boolean?,
    val categories: MutableMap<String, String>,
    val keywords: MutableList<String>,
    val neighbors: MutableList<String>,
    val parameters: MutableList<ContextParameter>
) {
    /**
     * @constructor Initialize builder for [WidgetContext]
     * @property url defines url for widget, required.
     * @property referrer defines referrer url.
     * @property pageclass the pageclass of the current page.
     * @property sentiment the sentiment of the current page.
     * @property recommending recs-recommending setting of the current page.
     * @property categories map for categories of the current page.
     * @property keywords list of keywords describing the context.
     * @property neighbors list of linked articles IDs.
     * @property parameters list of [ContextParameter] objects.
     */
    data class Builder @JvmOverloads constructor(
        var url: String,
        var referrer: String? = null,
        var pageclass: String? = null,
        var sentiment: String? = null,
        var recommending: Boolean? = null,
        var categories: MutableMap<String, String> = mutableMapOf(),
        var keywords: MutableList<String> = mutableListOf(),
        var neighbors: MutableList<String> = mutableListOf(),
        var parameters: MutableList<ContextParameter> = mutableListOf()
    ) {
        /**
         * Sets url
         * @param url url for widget.
         */
        fun url(url: String) = apply { this.url = url }

        /**
         * Sets referrer
         * @param referrer referrer url.
         */
        fun referrer(referrer: String?) = apply { this.referrer = referrer }

        /**
         * Sets pageclass
         * @param pageclass the pageclass of the current page.
         */
        fun pageclass(pageclass: String?) = apply { this.pageclass = pageclass }

        /**
         * Sets sentiment
         * @param sentiment the sentiment of the current page.
         */
        fun sentiment(sentiment: String?) = apply { this.sentiment = sentiment }

        /**
         * Sets recommending flag
         * @param recommending recs-recommending setting of the current page.
         */
        fun recommending(recommending: Boolean?) = apply { this.recommending = recommending }

        /**
         * Sets categories
         * @param categories map for categories of the current page.
         */
        fun categories(categories: MutableMap<String, String>) = apply { this.categories = categories }

        /**
         * Sets keywords
         * @param keywords list of keywords describing the context.
         */
        fun keywords(keywords: MutableList<String>) = apply { this.keywords = keywords }

        /**
         * Sets neighbors
         * @param neighbors list of linked articles IDs.
         */
        fun neighbors(neighbors: MutableList<String>) = apply { this.neighbors = neighbors }

        /**
         * Sets parameters
         * @param parameters list of [ContextParameter] objects.
         */
        fun parameters(parameters: MutableList<ContextParameter>) = apply { this.parameters = parameters }

        /**
         * Builds widget context
         * @throws [IllegalArgumentException] if constraints failed
         */
        fun build(): WidgetContext {
            check(url.toHttpUrlOrNull() != null) {
                "You should provide valid url as source"
            }
            referrer?.let {
                check(it.toHttpUrlOrNull() != null) {
                    "You should provide valid url as referrer"
                }
            }
            return WidgetContext(
                url,
                referrer,
                pageclass,
                sentiment,
                recommending,
                Collections.unmodifiableMap(categories),
                Collections.unmodifiableList(keywords),
                Collections.unmodifiableList(neighbors),
                Collections.unmodifiableList(parameters)
            )
        }
    }
}
